staging/src/k8s.io/apiextensions-apiserver/pkg/storageversion/manager.go

```diff
@@ -0,0 +1,297 @@
+		return nil
+	case <-ctx.Done():
+		return fmt.Errorf("aborted waiting for CRD storage version update: %w", ctx.Err())
+	// Unblock the requests if the storage version update takes a long time, otherwise
+	// CR requests may stack up and overwhelm the API server.
+	// TODO(roycaihw): benchmark the storage version update latency to adjust the timeout.
+	case <-time.After(storageVersionUpdateTimeout):
+		return fmt.Errorf("timeout waiting for CRD storage version update")
+	}
+
+}
+
+func (m *Manager) RunUpdateLoop(stopCh <-chan struct{}) {
+	go wait.Until(m.worker, time.Second, stopCh)
+	defer m.queue.ShutDown()
+	<-stopCh
+}
+
+func (m *Manager) worker() {
+	for m.processLatestSVUpdate() {
+	}
+}
+
+func (m *Manager) processLatestSVUpdate() bool {
+	ctx := context.TODO()
+	key, quit := m.queue.Get()
+	if quit {
+		return false
+	}
+	defer m.queue.Done(key)
+
+	val, ok := latestStorageVersionUpdateInfo.Load(key)
+	latestSVUpdateInfo := val.(*storageVersionUpdateInfo)
+	if !ok || latestSVUpdateInfo.processed {
+		klog.V(2).Infof("No pending storageversion update found, returning")
+		return true
+	}
+
+	m.updateStorageVersion(ctx, latestSVUpdateInfo.crd, latestSVUpdateInfo.updateChannels.teardownFinishedCh, latestSVUpdateInfo.updateChannels.processedCh, latestSVUpdateInfo.updateChannels.errCh)
+	markUpdateAsProcessed(latestSVUpdateInfo)
+	return true
+}
+
+func (m *Manager) recordLatestSVUpdateInfoWithTeardown(crd *apiextensionsv1.CustomResourceDefinition, tearDownFinishedCh <-chan struct{}) {
+	svUpdateInfo := &storageVersionUpdateInfo{
+		crd: crd,
+		updateChannels: &storageVersionUpdateChannels{
+			processedCh:        make(chan struct{}),
+			errCh:              make(chan struct{}),
+			teardownFinishedCh: tearDownFinishedCh,
+		},
+	}
+
+	// overwrite existing SVUpdateInfo with latest update event.
+	m.updateLatestStorageVersionUpdateInfo(crd.UID, svUpdateInfo)
+}
+
+func (m *Manager) updateLatestStorageVersionUpdateInfo(crdUID types.UID, svUpdateInfo *storageVersionUpdateInfo) {
+	latestStorageVersionUpdateInfo.Store(crdUID, svUpdateInfo)
+}
+
+// updateStorageVersion updates a StorageVersion for the given
+// CRD and returns immediately. Optionally, the caller may specify a
+// non-nil waitCh and/or a non-nil processedCh.
+// A non-nil waitCh will block the StorageVersion update until waitCh is
+// closed.
+// The manager will close the non-nil processedCh if it finished
+// processing the StorageVersion update (note that the update can either
+// succeeded or failed).
+func (m *Manager) updateStorageVersion(ctx context.Context, crd *apiextensionsv1.CustomResourceDefinition,
+	waitCh <-chan struct{}, processedCh chan<- struct{}, errCh chan<- struct{}) {
+	if waitCh != nil {
+		done := false
+		for {
+			select {
+			case <-waitCh:
+				done = true
+			case <-ctx.Done():
+				klog.V(4).Infof("aborting storageversion update for %v", crd)
+				if errCh != nil {
+					close(errCh)
+				}
+				return
+			case <-time.After(1 * time.Minute):
```

Comment:
Should this `1 * time.Minute` also be defined as a constant like `storageVersionUpdateTimeout` is?

Reply:
Yeah good point. Will incorporate this in the [main PR](https://github.com/kubernetes/kubernetes/pull/123999)

Reply:
Done in the [main PR](https://github.com/kubernetes/kubernetes/pull/123999)